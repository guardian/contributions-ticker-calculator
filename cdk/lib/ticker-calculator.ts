import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import type { GuScheduledLambdaProps } from '@guardian/cdk/lib/patterns/scheduled-lambda';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { RuleTargetInput, Schedule } from 'aws-cdk-lib/aws-events';
import { PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { LoggingFormat, Runtime } from 'aws-cdk-lib/aws-lambda';

const AppName = 'ticker-calculator';
const TickerBucket = 'contributions-ticker';

export class TickerCalculator extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const scheduleRules: GuScheduledLambdaProps['rules'] = [
			{
				schedule: Schedule.rate(Duration.minutes(15)),
				description: 'US',
				input: RuleTargetInput.fromText('US'),
			},
			{
				schedule: Schedule.rate(Duration.minutes(15)),
				description: 'AU',
				input: RuleTargetInput.fromText('AU'),
			},
		];

		const role = new Role(this, 'query-lambda-role', {
			// Set the name of the role rather than using an autogenerated name.
			// This is because if the ARN is too long then it breaks the authentication request to GCP
			roleName: `${AppName}-${this.stage}`,
			assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
		});
		role.addToPolicy(
			// Logging permissions
			new PolicyStatement({
				actions: [
					'logs:CreateLogGroup',
					'logs:CreateLogStream',
					'logs:PutLogEvents',
				],
				resources: ['*'],
			}),
		);
		role.addToPolicy(
			// Permission to create the ticker output file in S3
			new PolicyStatement({
				actions: ['s3:*'],
				resources: [
					`arn:aws:s3::*:${TickerBucket}/*`,
					`arn:aws:s3::*:${TickerBucket}`,
				],
			}),
		);
		role.addToPolicy(
			// Permission to read config from Parameter Store
			new PolicyStatement({
				actions: ['ssm:GetParameter'],
				resources: [
					`arn:aws:ssm:${this.region}:${this.account}:parameter/ticker-calculator/${this.stage}/gcp-wif-credentials-config`,
					`arn:aws:ssm:${this.region}:${this.account}:parameter/ticker-calculator/${this.stage}/ticker-config`,
				],
			}),
		);

		new GuScheduledLambda(this, 'TickerCalculator', {
			app: AppName,
			functionName: `${AppName}-${this.stage}`,
			runtime: Runtime.NODEJS_20_X,
			handler: 'lambda.handler',
			fileName: `${AppName}.zip`,
			rules: scheduleRules,
			role,
			environment: {
				Stage: this.stage,
			},
			timeout: Duration.minutes(2),
			loggingFormat: LoggingFormat.TEXT,
			monitoringConfiguration:
				this.stage === 'PROD'
					? {
							toleratedErrorPercentage: 0,
							snsTopicName: 'alarms-handler-topic-PROD',
						}
					: { noMonitoring: true },
		});
	}
}
