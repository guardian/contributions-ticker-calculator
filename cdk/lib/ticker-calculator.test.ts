import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { TickerCalculator } from './ticker-calculator';

describe('The TickerCalculator stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new TickerCalculator(app, 'TickerCalculator', {
			stack: 'support',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
