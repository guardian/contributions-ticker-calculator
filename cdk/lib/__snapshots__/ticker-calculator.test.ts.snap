// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The TickerCalculator stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TickerCalculator5AB41211": {
      "DependsOn": [
        "querylambdaroleDefaultPolicy929B9FB7",
        "querylambdaroleC6C8A94B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/TEST/ticker-calculator/ticker-calculator.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "ticker-calculator",
            "STACK": "support",
            "STAGE": "TEST",
            "Stage": "TEST",
          },
        },
        "FunctionName": "ticker-calculator-TEST",
        "Handler": "lambda.handler",
        "LoggingConfig": {
          "LogFormat": "Text",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "querylambdaroleC6C8A94B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "ticker-calculator",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/contributions-ticker-calculator",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TickerCalculatorTickerCalculatorrate15minutes0A13A653E": {
      "Properties": {
        "Description": "US",
        "ScheduleExpression": "rate(15 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "TickerCalculator5AB41211",
                "Arn",
              ],
            },
            "Id": "Target0",
            "Input": ""US"",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TickerCalculatorTickerCalculatorrate15minutes0AllowEventRuleTickerCalculator3A61D8F5E0A4FFD1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TickerCalculator5AB41211",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "TickerCalculatorTickerCalculatorrate15minutes0A13A653E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TickerCalculatorTickerCalculatorrate15minutes161311C65": {
      "Properties": {
        "Description": "AU",
        "ScheduleExpression": "rate(15 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "TickerCalculator5AB41211",
                "Arn",
              ],
            },
            "Id": "Target0",
            "Input": ""AU"",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TickerCalculatorTickerCalculatorrate15minutes1AllowEventRuleTickerCalculator3A61D8F5E506B789": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TickerCalculator5AB41211",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "TickerCalculatorTickerCalculatorrate15minutes161311C65",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TickerCalculatorTickerCalculatorrate15minutes2AllowEventRuleTickerCalculator3A61D8F5B129E114": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TickerCalculator5AB41211",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "TickerCalculatorTickerCalculatorrate15minutes2F658BDF5",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TickerCalculatorTickerCalculatorrate15minutes2F658BDF5": {
      "Properties": {
        "Description": "global",
        "ScheduleExpression": "rate(15 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "TickerCalculator5AB41211",
                "Arn",
              ],
            },
            "Id": "Target0",
            "Input": ""global"",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "querylambdaroleC6C8A94B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "ticker-calculator-TEST",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/contributions-ticker-calculator",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "querylambdaroleDefaultPolicy929B9FB7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3:*",
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3::*:contributions-ticker/*",
                "arn:aws:s3::*:contributions-ticker",
              ],
            },
            {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:ssm:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":parameter/ticker-calculator/TEST/gcp-wif-credentials-config",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:ssm:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":parameter/ticker-calculator/TEST/ticker-config",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/TEST/ticker-calculator/ticker-calculator.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/support/ticker-calculator",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/support/ticker-calculator/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "querylambdaroleDefaultPolicy929B9FB7",
        "Roles": [
          {
            "Ref": "querylambdaroleC6C8A94B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
